src/Bridge/Symfony/Validator/Metadata/Property/Restriction/PropertySchemaCollectionRestriction.php

```diff
@@ -0,0 +1,128 @@
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Bridge\Symfony\Validator\Metadata\Property\Restriction;
+
+use ApiPlatform\Core\Metadata\Property\PropertyMetadata;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\Constraints\Collection;
+use Symfony\Component\Validator\Constraints\Optional;
+use Symfony\Component\Validator\Constraints\Required;
+use Symfony\Component\Validator\Constraints\Type;
+
+/**
+ * @author Tomas Norkūnas <norkunas.tom@gmail.com>
+ */
+final class PropertySchemaCollectionRestriction implements PropertySchemaRestrictionMetadataInterface
+{
+    /**
+     * @var iterable<PropertySchemaRestrictionMetadataInterface>
+     */
+    private $restrictionsMetadata;
+
+    /**
+     * @param iterable<PropertySchemaRestrictionMetadataInterface> $restrictionsMetadata
+     */
+    public function __construct(iterable $restrictionsMetadata = [])
+    {
+        $this->restrictionsMetadata = $restrictionsMetadata;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @param Collection $constraint
+     */
+    public function create(Constraint $constraint, PropertyMetadata $propertyMetadata): array
+    {
+        $restriction = [
+            'type' => 'object',
+            'properties' => [],
+            'additionalProperties' => $constraint->allowExtraFields,
+        ];
+        $required = [];
+
+        foreach ($constraint->fields as $field => $baseConstraint) {
+            /** @var Required|Optional $baseConstraint */
+            if ($baseConstraint instanceof Required) {
+                $required[] = $field;
+            }
+
+            $restriction['properties'][$field] = $this->mergeConstraintRestrictions($baseConstraint, $propertyMetadata);
+        }
+
+        if ($required) {
+            $restriction['required'] = $required;
+        }
+
+        return $restriction;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supports(Constraint $constraint, PropertyMetadata $propertyMetadata): bool
+    {
+        return $constraint instanceof Collection;
+    }
+
+    /**
+     * @param Required|Optional $constraint
+     */
+    private function mergeConstraintRestrictions(Constraint $constraint, PropertyMetadata $propertyMetadata): array
+    {
+        $propertyRestrictions = [['type' => 'string']];
+        $nestedConstraints = method_exists($constraint, 'getNestedContraints') ? $constraint->getNestedContraints() : $constraint->constraints;
+
+        foreach ($nestedConstraints as $nestedConstraint) {
+            if ($nestedConstraint instanceof Type) {
+                /** @var string $nestedType */
+                $nestedType = $nestedConstraint->type;
+                $propertyRestrictions[0]['type'] = $this->getConstraintType($nestedType);
+            }
```

Comment:
Should we really add the type? We don't add it for a property, isn't it?

Reply:
trully not sure :)

Reply:
if the property metadata would have types for nested fields then we probably shouldn't, but when i declare a nested field which must be an integer i'd expect to that the int would be exposed

Reply:
I'm not sure to understand. Today, the `type` restriction is not added to the schema for a top-level property. Why would it be defined for a nested one?

Reply:
As you resist from this I've removed the type guessing :) I agree that this code should not be responsible for this, but then I think api-platform should ensure that it'd at least possible to do this:
```
/**
 * @var array{age:int, social:array{githubUsername:?string}}
 * @Assert\Collection(
 *     fields={
 *         "age"=..,
 *         "social"=@Assert\Collection({
 *            fields={"githubUsername"=..}
 *         })
 *     }
 * )
 */
```
and the types would be exposed from typedef

Reply:
I'm not saying you are wrong :slightly_smiling_face: It's just that we need to be coherent. We should add the type to the top-level properties beforehand if we want to add it to the nested ones.
I'm not sure why there is no type for the top-level properties. Anyway it's not the `Validator` responsibility to add the types, if we add it, it should be elsewhere.

Reply:
I think that for top-level properties types are fetched with property-info component, but not sure if it can provide nested ones

Reply:
But they are not added to the `schema`, aren't they?

Reply:
could be :)